File: Python/ByRound/2001/2001_E1_Deterministic_Heap_Easy_Version.py
# Problem: CF 2001 E1 - Deterministic Heap (Easy Version)
# https://codeforces.com/contest/2001/problem/E1

"""
Algorithm: Dynamic Programming with Prefix Sums and Modular Arithmetic
Time Complexity: O(n * k^2)
Space Complexity: O(n * k)

This solution counts the number of deterministic max-heaps that can be formed
by performing exactly k "add" operations on a perfect binary tree of height n.
Each operation adds 1 to all nodes on the path from root to a chosen node.

Key techniques used:
1. Dynamic Programming: dp[i][j] represents number of ways to form valid heaps
   at height i with exactly j operations.
2. Prefix Sums: Efficient computation of sums in the recurrence relation.
3. Modular Arithmetic: All operations done modulo p for large numbers.
4. Inverse Elements: Precomputed modular inverses for division in modular arithmetic.

The algorithm uses the recurrence relation based on how operations are distributed
across subtree levels, with careful handling of the deterministic heap constraint.
"""

import os
import sys
from io import BytesIO, IOBase


def main():
    t = int(input())
    for _ in range(t):
        n, k, p = map(int, input().split())
        # Precompute modular inverses for later use
        inv = [-1] + [pow(i, -1, p) for i in range(1, k + 1)]
        # dp[i][j] = number of ways to get exactly j operations at height i
        dp = [[0] * (k + 1) for _ in range(n + 1)]
        # Base case: at height 1, we can only have 0 operations (empty heap) or any number of operations
        dp[1] = [1] * (k + 1)
        dp[1][0] = 0  # We must perform at least one operation, so dp[1][0] is 0
        for i in range(2, n + 1):
            # pref[x] = sum of choose from 0 to x (prefix sum)
            pref, choose, child_sz = [0] * k, 1, (pow(2, i - 1, p) - 1) % p
            # Compute prefix sums for combinations using recurrence
            for x in range(k):
                pref[x] = (pref[x - 1] + choose) % p
                choose = (choose * (child_sz + x) * inv[x + 1]) % p
            for j in range(1, k + 1):
                # For each possible split of operations at current level
                for hi in range(1, j + 1):
                    lo = min(hi - 1, j - hi)
                    # Add to the result the contribution from current split
                    dp[i][j] = (dp[i][j] + 2 * pref[lo] * dp[i - 1][hi]) % p
        print(dp[n][k])


BUFSIZE = 8192


class FastIO(IOBase):
    newlines = 0

    def __init__(self, file):
        self._file = file
        self._fd = file.fileno()
        self.buffer = BytesIO()
        self.writable = "x" in file.mode or "r" not in file.mode
        self.write = self.buffer.write if self.writable else None

    def read(self):
        while True:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            if not b:
                break
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines = 0
        return self.buffer.read()

    def readline(self):
        while self.newlines == 0:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            self.newlines = b.count(b"\n") + (not b)
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines -= 1
        return self.buffer.readline()

    def flush(self):
        if self.writable:
            os.write(self._fd, self.buffer.getvalue())
            self.buffer.truncate(0), self.buffer.seek(0)


class IOWrapper(IOBase):
    def __init__(self, file):
        self.buffer = FastIO(file)
        self.flush = self.buffer.flush
        self.writable = self.buffer.writable
        self.write = lambda s: self.buffer.write(s.encode("ascii"))
        self.read = lambda: self.buffer.read().decode("ascii")
        self.readline = lambda: self.buffer.readline().decode("ascii")


sys.stdin, sys.stdout = IOWrapper(sys.stdin), IOWrapper(sys.stdout)
input = lambda: sys.stdin.readline().rstrip("\r\n")
if __name__ == "__main__":
    main()


# https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/E/2001_E1_Deterministic_Heap_Easy_Version.cpp
// Problem: CF 2001 E1 - Deterministic Heap (Easy Version)
// https://codeforces.com/contest/2001/problem/E1

/*
 * Problem: E1. Deterministic Heap (Easy Version)
 * Purpose: Count the number of different deterministic max-heaps that can be formed by applying 'add' operation exactly k times on a perfect binary tree of size 2^n - 1.
 * Algorithms/Techniques:
 *   - Dynamic Programming with memoization
 *   - Tree traversal logic and heap property constraints
 * Time Complexity: O(n * k^2) per test case
 * Space Complexity: O(n * k) for DP tables
 */

#include <stdint.h>
#include <string.h>
#include <iostream>

using namespace std;
const int MAX = 505;
int dp_f[MAX][MAX], dp_g[MAX][MAX];
int MOD;

// Function g(n, k): Computes number of ways to distribute k operations among n nodes such that heap properties hold
int g(int n, int k) {
  if (k < 0)
    return 0;
  if (n == 1)
    return 1;
  int &ret = dp_g[n][k];
  if (~ret)
    return ret;
  ret = g(n, k - 1); // Base case: do not use any operation
  for (int a = 0; a <= k; a++) {
    int b = k - a;
    ret = (ret + 1LL * g(n - 1, a) * g(n - 1, b)) % MOD;
  }
  return ret;
}

// Function f(n, k): Computes number of deterministic max-heaps with n nodes and k operations
int f(int n, int k) {
  if (k < 0)
    return 0;
  if (n == 1)
    return 1;
  int &ret = dp_f[n][k];
  if (~ret)
    return ret;
  ret = f(n, k - 1); // Base case: do not use any operation
  for (int a = 0; 2 * a < k; a++) {
    int b = k - a;
    ret = (ret + 2LL * g(n - 1, a) * f(n - 1, b)) % MOD;
  }
  return ret;
}

void solve() {
  int n, k;
  cin >> n >> k >> MOD;
  memset(dp_f, -1, sizeof dp_f);
  memset(dp_g, -1, sizeof dp_g);
  cout << f(n, k) << '\n';
}

int32_t main() {
  ios_base::sync_with_stdio(0), cin.tie(0);
  int t = 1;
  cin >> t;
  while (t--)
    solve();
}


// https://github.com/VaHiX/codeForces/